<a href="https://colab.research.google.com/github/johnnysDeu/Thesis/blob/main/Model_DataAugmentation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import keras
keras.__version__

'2.15.0'

In [ ]:
import os, shutil
import numpy as np

In [ ]:
from google.colab import drive # fur colab
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
#!nvidia-smi

In [ ]:
path = '/content/drive/MyDrive/Colab Notebooks/Thesis/data'
os.chdir(path)

In [ ]:
Current_dir = os.getcwd()
Current_dir

'/content/drive/MyDrive/Colab Notebooks/Thesis/data'

In [ ]:
base_dir = 'data split'
#os.mkdir(base_dir)

# Directories for our training,
# validation and test splits
train_dir = os.path.join(base_dir, 'train')
if not os.path.isdir(train_dir):
  os.mkdir(train_dir)
validation_dir = os.path.join(base_dir, 'validate')
if not os.path.isdir(validation_dir):
  os.mkdir(validation_dir)
test_dir = os.path.join(base_dir, 'test')
if not os.path.isdir(test_dir):
  os.mkdir(test_dir)


In [ ]:
train_Ads_dir = os.path.join(train_dir, 'Ads')
train_sample_dir = os.path.join(train_dir, 'Sample_images')
train_Ads_dir

'data split/train/Ads'

In [ ]:
validation_Ads_dir = os.path.join(validation_dir, 'Ads')
validation_sample_dir = os.path.join(validation_dir, 'Sample_images')
validation_Ads_dir

'data split/validate/Ads'

In [ ]:
test_Ads_dir = os.path.join(test_dir, 'Ads')
test_sample_dir = os.path.join(test_dir, 'Sample_images')
test_Ads_dir

'data split/test/Ads'

Block comment : crtl + # <br>
F11 run next

In [ ]:
# The path to the directory where the original
# dataset was uncompressed
#original_dataset_dir = '/Users/fchollet/Downloads/kaggle_original_data'
#original_dataset_dir = 'C:/Users/YannisPC/Desktop/Courses/MSc/7.DEEP NEURAL NETWORKS/Labs/lab8/trainInitial'

# The directory where we will
# store our smaller dataset
#base_dir = '/Users/fchollet/Downloads/cats_and_dogs_small'

# Directory with our training cat pictures
train_Ads_dir = os.path.join(train_dir, 'Ads')
if not os.path.isdir(train_Ads_dir):
    os.mkdir(train_Ads_dir)

# Directory with our training dog pictures
train_sample_dir = os.path.join(train_dir, 'Sample_images')
if not os.path.isdir(train_sample_dir):
    os.mkdir(train_sample_dir)

# Directory with our validation cat pictures
validation_Ads_dir = os.path.join(validation_dir, 'Ads')
if not os.path.isdir(validation_Ads_dir):
    os.mkdir(validation_Ads_dir)

# Directory with our validation dog pictures
validation_sample_dir = os.path.join(validation_dir, 'Sample_images')
if not os.path.isdir(validation_sample_dir):
    os.mkdir(validation_sample_dir)

# Directory with our validation cat pictures
test_Ads_dir = os.path.join(test_dir, 'Ads')
if not os.path.isdir(test_Ads_dir):
    os.mkdir(test_Ads_dir)

# Directory with our validation dog pictures
test_sample_dir = os.path.join(test_dir, 'Sample_images')
if not os.path.isdir(test_sample_dir):
    os.mkdir(test_sample_dir)



In [ ]:
# original_dataset_dir_imges = r'C:\Users\YannisPC\PycharmProjects\Thesis\Thesis\Sample_Images_10000'
# original_dataset_dir_Ads = r'C:\Users\YannisPC\PycharmProjects\Thesis\Thesis\Ads'


# files = os.listdir(original_dataset_dir_imges)
# image_files = [f for f in files if f.lower().endswith(('.jpg', '.jpeg'))]
# #image_files

# file_ads = os.listdir(original_dataset_dir_Ads)
# ads_files = [f for f in file_ads if f.lower().endswith(('.jpg', '.jpeg', '.png'))]

In [ ]:
# import random
# import shutil

# random.shuffle(image_files)
# random.shuffle(ads_files)

1200 simple images <br> 1200 Ads

0.6 Train, 0.2 Val, 0.2 Test

In [ ]:
# # Copy first 720 simple images to train_sample_dir
# for image_file in image_files[:720]:
#     source_path = os.path.join(original_dataset_dir_imges, image_file)
#     destination_path = os.path.join(train_sample_dir, image_file)
#     shutil.copyfile(source_path, destination_path)
#     #print(f"Copied {image_file} to {train_sample_dir}")

In [ ]:
# # Copy first 720 Ad images to train_Ads_dir

# for ad_file in ads_files[:720]:
#     source_path = os.path.join(original_dataset_dir_Ads, ad_file)
#     destination_path = os.path.join(train_Ads_dir, ad_file)
#     shutil.copyfile(source_path, destination_path)
#     #print(f"Copied {ad_file} to {train_Ads_dir}")


In [ ]:
# # Copy next 240 simple images to val_sample_dir
# for image_file in image_files[720:960]:
#     source_path = os.path.join(original_dataset_dir_imges, image_file)
#     destination_path = os.path.join(validation_sample_dir, image_file)
#     shutil.copyfile(source_path, destination_path)
#     #print(f"Copied {image_file} to {validation_sample_dir}")

In [ ]:
# # Copy next 240 Ad images to val_Ads_dir

# for ad_file in ads_files[720:960]:
#     source_path = os.path.join(original_dataset_dir_Ads, ad_file)
#     destination_path = os.path.join(validation_Ads_dir, ad_file)
#     shutil.copyfile(source_path, destination_path)
#     #print(f"Copied {ad_file} to {validation_Ads_dir}")


In [ ]:
# # Copy final 240 simple images to test_sample_dir
# for image_file in image_files[960:1200]:
#     source_path = os.path.join(original_dataset_dir_imges, image_file)
#     destination_path = os.path.join(test_sample_dir, image_file)
#     shutil.copyfile(source_path, destination_path)
#     #print(f"Copied {image_file} to {test_sample_dir}")

In [ ]:
# # Copy Final 240 Ad images to test_Ads_dir

# for ad_file in ads_files[960:1200]:
#     source_path = os.path.join(original_dataset_dir_Ads, ad_file)
#     destination_path = os.path.join(test_Ads_dir, ad_file)
#     shutil.copyfile(source_path, destination_path)
#     #print(f"Copied {ad_file} to {test_Ads_dir}")

As a sanity check, let's count how many pictures we have in each training split (train/validation/test):

In [ ]:
print('total training Ads images:', len(os.listdir(train_Ads_dir)))

total training Ads images: 719


In [ ]:
print('total training sample images:', len(os.listdir(train_sample_dir)))

total training sample images: 720


In [ ]:
print('total validation Ads images:', len(os.listdir(validation_Ads_dir)))

total validation Ads images: 226


In [ ]:
print('total validation sample images:', len(os.listdir(validation_sample_dir)))

total validation sample images: 234


In [ ]:
print('total test Ads images:', len(os.listdir(test_Ads_dir)))

total test Ads images: 240


In [ ]:
print('total test sample images:', len(os.listdir(test_sample_dir)))

total test sample images: 260


In [ ]:
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Dense, Flatten ,Dropout, BatchNormalization

from keras import layers
from keras import models

model = models.Sequential()
model.add(layers.Conv2D(32, (3, 3), activation='relu',
                        input_shape=(150, 150, 3)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(128, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(128, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Flatten())
model.add(layers.Dropout(0.2))
model.add(layers.Dense(512, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

In [ ]:
from keras import layers
from keras import models

model = models.Sequential()
model.add(layers.Conv2D(32, (3,3), 1, activation='relu', input_shape=(150,150,3))) # mhpws 150,150
model.add(layers.MaxPooling2D())
model.add(BatchNormalization())

model.add(layers.Conv2D(64, (3,3), 1, activation='relu'))
model.add(layers.MaxPooling2D())
model.add(BatchNormalization())

model.add(layers.Conv2D(128, (3,3), 1, activation='relu'))
model.add(layers.MaxPooling2D())
model.add(BatchNormalization())

model.add(layers.Conv2D(128, (3,3), 1, activation='relu'))
model.add(layers.MaxPooling2D())
model.add(layers.Dropout(0.2))

model.add(layers.Flatten())
model.add(layers.Dense(256, activation='relu'))
model.add(BatchNormalization())

model.add(layers.Dense(1, activation='sigmoid'))

Let's take a look at how the dimensions of the feature maps change with every successive layer:

In [ ]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 148, 148, 32)      896       
                                                                 
 max_pooling2d (MaxPooling2  (None, 74, 74, 32)        0         
 D)                                                              
                                                                 
 batch_normalization (Batch  (None, 74, 74, 32)        128       
 Normalization)                                                  
                                                                 
 conv2d_1 (Conv2D)           (None, 72, 72, 64)        18496     
                                                                 
 max_pooling2d_1 (MaxPoolin  (None, 36, 36, 64)        0         
 g2D)                                                            
                                                        

In [ ]:
from keras import optimizers

model.compile(loss='binary_crossentropy',
              optimizer=optimizers.RMSprop(learning_rate=1e-4),
              metrics=['acc'])

## Data preprocessing

As you already know by now, data should be formatted into appropriately pre-processed floating point tensors before being fed into our
network. Currently, our data sits on a drive as JPEG files, so the steps for getting it into our network are roughly:

* Read the picture files.
* Decode the JPEG content to RBG grids of pixels.
* Convert these into floating point tensors.
* Rescale the pixel values (between 0 and 255) to the [0, 1] interval (as you know, neural networks prefer to deal with small input values).

It may seem a bit daunting, but thankfully Keras has utilities to take care of these steps automatically. Keras has a module with image
processing helper tools, located at `keras.preprocessing.image`. In particular, it contains the class `ImageDataGenerator` which allows to
quickly set up Python generators that can automatically turn image files on disk into batches of pre-processed tensors. This is what we
will use here.

In [ ]:
from keras.preprocessing.image import ImageDataGenerator

# All images will be rescaled by 1./255
train_datagen = ImageDataGenerator(rescale=1./255)
test_datagen = ImageDataGenerator(rescale=1./255)

train_generator = train_datagen.flow_from_directory(
        # This is the target directory
        train_dir,
        # All images will be resized to 150x150
        target_size=(150, 150),
        batch_size=32,
        # Since we use binary_crossentropy loss, we need binary labels
        class_mode='binary')

validation_generator = test_datagen.flow_from_directory(
        validation_dir,
        target_size=(150, 150),
        batch_size=32,
        class_mode='binary')


batch_size = 32

Found 1439 images belonging to 2 classes.
Found 460 images belonging to 2 classes.


Let's take a look at the output of one of these generators: it yields batches of 150x150 RGB images (shape `(20, 150, 150, 3)`) and binary
labels (shape `(20,)`). 20 is the number of samples in each batch (the batch size). Note that the generator yields these batches
indefinitely: it just loops endlessly over the images present in the target folder. For this reason, we need to `break` the iteration loop
at some point.

In [ ]:
cnt = 0
for data_batch, labels_batch in train_generator:
    print('data batch shape:', data_batch.shape)
    print('labels batch shape:', labels_batch.shape)
    break

data batch shape: (32, 150, 150, 3)
labels batch shape: (32,)


Let's fit our model to the data using the generator. We do it using the `fit_generator` method, the equivalent of `fit` for data generators
like ours. It expects as first argument a Python generator that will yield batches of inputs and targets indefinitely, like ours does.
Because the data is being generated endlessly, the generator needs to know example how many samples to draw from the generator before
declaring an epoch over. This is the role of the `steps_per_epoch` argument: after having drawn `steps_per_epoch` batches from the
generator, i.e. after having run for `steps_per_epoch` gradient descent steps, the fitting process will go to the next epoch. In our case,
batches are 20-sample large, so it will take 100 batches until we see our target of 2000 samples.

When using `fit_generator`, one may pass a `validation_data` argument, much like with the `fit` method. Importantly, this argument is
allowed to be a data generator itself, but it could be a tuple of Numpy arrays as well. If you pass a generator as `validation_data`, then
this generator is expected to yield batches of validation data endlessly, and thus you should also specify the `validation_steps` argument,
which tells the process how many batches to draw from the validation generator for evaluation.

Naturally what you want is in 1 epoch your generator to pass through all of your training data one time

steps_per_epoch = int( np.ceil(x_train.shape[0] / batch_size) )

In [ ]:
len(train_generator)
len(validation_generator)

epochs_len = len(train_generator)
Val_len = len(validation_generator)
print("Epochs len: ", epochs_len)
print("Val len: ", Val_len)



Epochs len:  45
Val len:  15


`steps_per_epoch * epochs` batches (in this case, 1500 batches).  

steps_per_epoch= 50\\
epochs = 30

training set = 784 \\
val set = 316

training set = 1440 \\
val set = 420


In [ ]:
from keras.callbacks import ReduceLROnPlateau

In [ ]:
reduce_lr = ReduceLROnPlateau(monitor='val_loss', factor=0.2, verbose = 1,
                              patience=3, min_lr=0.001)

In [ ]:
history = model.fit(    #fit_generator
      train_generator,
      steps_per_epoch=epochs_len,
      epochs=30,
      validation_data=validation_generator,
      validation_steps=Val_len,
      callbacks=[reduce_lr])

Epoch 1/30
45/45 [==============================] - 339s 8s/step - loss: 0.4750 - acc: 0.7929 - val_loss: 0.6945 - val_acc: 0.4739 - lr: 1.0000e-04


KeyboardInterrupt: 

It is good practice to always save your models after training:

In [ ]:
#model.save('Ads_classifier_small_1.h5')

In [ ]:
START_PLOT_FROM_EPOCH = 3

Let's plot the loss and accuracy of the model over the training and validation data during training:

In [ ]:
import matplotlib.pyplot as plt

acc = history.history['acc'][START_PLOT_FROM_EPOCH:]
val_acc = history.history['val_acc'][START_PLOT_FROM_EPOCH:]
loss = history.history['loss'][START_PLOT_FROM_EPOCH:]
val_loss = history.history['val_loss'][START_PLOT_FROM_EPOCH:]

epochs = range(len(acc))

plt.plot(epochs, acc, 'bo', label='Training acc')
plt.plot(epochs, val_acc, 'b', label='Validation acc')
plt.title('Training and validation accuracy')
plt.legend()

plt.figure()

plt.plot(epochs, loss, 'bo', label='Training loss')
plt.plot(epochs, val_loss, 'b', label='Validation loss')
plt.title('Training and validation loss')
plt.legend()

plt.show()

check this tread: \\
https://stackoverflow.com/questions/61568463/keras-validation-accuracy-stays-the-exact-same-but-validation-loss-decreases

We can now finally evaluate this model on the test data:

In [ ]:
test_generator = test_datagen.flow_from_directory(
        test_dir,
        target_size=(150, 150),
        batch_size=32,
        class_mode='binary')

test_loss, test_acc = model.evaluate_generator(test_generator, steps=len(test_generator))
print('test acc:', test_acc)
print('test loss:', test_loss)

These plots are characteristic of overfitting. Our training accuracy increases linearly over time, until it reaches nearly 100%, while our
validation accuracy stalls at 70-72%. Our validation loss reaches its minimum after only five epochs then stalls, while the training loss
keeps decreasing linearly until it reaches nearly 0.

Because we only have relatively few training samples (2000), overfitting is going to be our number one concern. You already know about a
number of techniques that can help mitigate overfitting, such as dropout and weight decay (L2 regularization). We are now going to
introduce a new one, specific to computer vision, and used almost universally when processing images with deep learning models: *data
augmentation*.

## Using data augmentation

Overfitting is caused by having too few samples to learn from, rendering us unable to train a model able to generalize to new data.
Given infinite data, our model would be exposed to every possible aspect of the data distribution at hand: we would never overfit. Data
augmentation takes the approach of generating more training data from existing training samples, by "augmenting" the samples via a number
of random transformations that yield believable-looking images. The goal is that at training time, our model would never see the exact same
picture twice. This helps the model get exposed to more aspects of the data and generalize better.

In Keras, this can be done by configuring a number of random transformations to be performed on the images read by our `ImageDataGenerator`
instance. Let's get started with an example:

In [ ]:
datagen = ImageDataGenerator(
      rotation_range=40,
      width_shift_range=0.2,
      height_shift_range=0.2,
      shear_range=0.2,
      zoom_range=0.2,
      horizontal_flip=True,
      fill_mode='nearest')

These are just a few of the options available (for more, see the Keras documentation). Let's quickly go over what we just wrote:

* `rotation_range` is a value in degrees (0-180), a range within which to randomly rotate pictures.
* `width_shift` and `height_shift` are ranges (as a fraction of total width or height) within which to randomly translate pictures
vertically or horizontally.
* `shear_range` is for randomly applying shearing transformations.
* `zoom_range` is for randomly zooming inside pictures.
* `horizontal_flip` is for randomly flipping half of the images horizontally -- relevant when there are no assumptions of horizontal
asymmetry (e.g. real-world pictures).
* `fill_mode` is the strategy used for filling in newly created pixels, which can appear after a rotation or a width/height shift.

Let's take a look at our augmented images:

In [ ]:
# This is module with image preprocessing utilities
#from keras.preprocessing import image
import keras.utils as image

fnames = [os.path.join(train_Ads_dir, fname) for fname in os.listdir(train_Ads_dir)]

# We pick one image to "augment"
img_path = fnames[4]

# Read the image and resize it
img = image.load_img(img_path, target_size=(150, 150))

# Convert it to a Numpy array with shape (150, 150, 3)
x = image.img_to_array(img)

# Reshape it to (1, 150, 150, 3)
x = x.reshape((1,) + x.shape)

# The .flow() command below generates batches of randomly transformed images.
# It will loop indefinitely, so we need to `break` the loop at some point!
i = 0
for batch in datagen.flow(x, batch_size=1):
    plt.figure(i)
    imgplot = plt.imshow(image.array_to_img(batch[0]))
    i += 1
    if i % 4 == 0:
        break

plt.show()

If we train a new network using this data augmentation configuration, our network will never see twice the same input. However, the inputs
that it sees are still heavily intercorrelated, since they come from a small number of original images -- we cannot produce new information,
we can only remix existing information. As such, this might not be quite enough to completely get rid of overfitting. To further fight
overfitting, we will also add a Dropout layer to our model, right before the densely-connected classifier:

In [ ]:
model_aug = models.Sequential()
model_aug.add(layers.Conv2D(32, (3, 3), activation='relu',
                        input_shape=(150, 150, 3)))
model_aug.add(layers.MaxPooling2D((2, 2)))
model_aug.add(layers.Conv2D(64, (3, 3), activation='relu'))
model_aug.add(layers.MaxPooling2D((2, 2)))
model_aug.add(layers.Conv2D(128, (3, 3), activation='relu'))
model_aug.add(layers.MaxPooling2D((2, 2)))
model_aug.add(layers.Conv2D(128, (3, 3), activation='relu'))
model_aug.add(layers.MaxPooling2D((2, 2)))
model_aug.add(layers.Flatten())
model_aug.add(layers.Dropout(0.5))
model_aug.add(layers.Dense(256, activation='relu'))
model_aug.add(layers.Dense(1, activation='sigmoid'))

model_aug.compile(loss='binary_crossentropy',
              optimizer=optimizers.RMSprop(learning_rate=1e-5),
              metrics=['acc'])

In [ ]:
len(validation_dir)

Let's train our network using data augmentation and dropout:

In [ ]:
train_datagen = ImageDataGenerator(
    rescale=1./255,
    rotation_range=40,
    width_shift_range=0.2,
    height_shift_range=0.2,
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True,)

# Note that the validation data should not be augmented!
test_datagen = ImageDataGenerator(rescale=1./255)

train_generator = train_datagen.flow_from_directory(
        # This is the target directory
        train_dir,
        # All images will be resized to 150x150
        target_size=(150, 150),
        batch_size=32, #32
        # Since we use binary_crossentropy loss, we need binary labels
        class_mode='binary')

validation_generator = test_datagen.flow_from_directory(
        validation_dir,
        target_size=(150, 150),
        batch_size=32,
        class_mode='binary')

# history = model.fit( #fit_generator
#       train_generator,
#       steps_per_epoch=50,
#       epochs=100,
#       validation_data=validation_generator,
#       validation_steps=50)


In [ ]:
history = model_aug.fit(    #fit_generator
      train_generator,
      steps_per_epoch=epochs_len,
      epochs=30,
      validation_data=validation_generator,
      validation_steps=Val_len)

Let's save our model -- we will be using it in the section on convnet visualization.

In [ ]:
#model.save('Ads_classifier_small_2.h5')

Let's plot our results again:

In [ ]:
acc = history.history['acc'][START_PLOT_FROM_EPOCH:]
val_acc = history.history['val_acc'][START_PLOT_FROM_EPOCH:]
loss = history.history['loss'][START_PLOT_FROM_EPOCH:]
val_loss = history.history['val_loss'][START_PLOT_FROM_EPOCH:]

epochs = range(len(acc))

plt.plot(epochs, acc, 'bo', label='Training acc')
plt.plot(epochs, val_acc, 'b', label='Validation acc')
plt.title('Training and validation accuracy')
plt.legend()

plt.figure()

plt.plot(epochs, loss, 'bo', label='Training loss')
plt.plot(epochs, val_loss, 'b', label='Validation loss')
plt.title('Training and validation loss')
plt.legend()

plt.show()

In [ ]:
test_generator = test_datagen.flow_from_directory(
        test_dir,
        target_size=(150, 150),
        batch_size=32,
        class_mode='binary')

test_loss, test_acc = model_aug.evaluate_generator(test_generator, steps=len(test_generator))  # steps_per_epoch * epochs
print('test acc:', test_acc)
print('test loss:', test_loss)

Thanks to data augmentation and dropout, we are no longer overfitting: the training curves are rather closely tracking the validation
curves. We are now able to reach an accuracy of 82%, a 15% relative improvement over the non-regularized model.

By leveraging regularization techniques even further and by tuning the network's parameters (such as the number of filters per convolution
layer, or the number of layers in the network), we may be able to get an even better accuracy, likely up to 86-87%. However, it would prove
very difficult to go any higher just by training our own convnet from scratch, simply because we have so little data to work with. As a
next step to improve our accuracy on this problem, we will have to leverage a pre-trained model, which will be the focus of the next two
sections.

end